flfm/app/main.py

```diff
@@ -0,0 +1,470 @@
+                                                    id="radius",
+                                                    debounce=app_settings.DEBOUNCE,
+                                                ),
+                                            ]
+                                        )
+                                    ),
+                                    dbc.Col(
+                                        html.Div(
+                                            [
+                                                dbc.Button("Reconstruct", id="run-reconstruction", n_clicks=0),
+                                            ]
+                                        )
+                                    ),
+                                ]
+                            ),
+                            dcc.Loading(
+                                id=dict(type="loading", index="reconstruction"),
+                                type=app_settings.LOADING_TYPE,
+                                delay_show=app_settings.LOADING_DELAY_SHOW,
+                                children=html.Div(id=dict(type="display", index="reconstruction")),
+                            ),
+                        ],
+                    ),
+                ]
+            ),
+        ]
+    )
+    return layout
+
+
+dash_app.layout = dash_layout
+
+
+@callback(
+    Output(dict(type="display", index=MATCH), "children", allow_duplicate=True),
+    Output(dict(type="n-frames", index=MATCH), "data"),
+    Input(dict(type="upload", index=MATCH), "contents"),
+    State(dict(type="upload", index=MATCH), "filename"),
+    State(dict(type="upload", index=MATCH), "id"),
+    prevent_initial_call=True,
+)
+def upload_data(contents, filename, id):
+    """Upload data/image files."""
+    global image_data
+    id = id["index"]
+    image_data[id] = upload_file(contents, filename)
+    children, n_frames_from_image = _display_image_with_slider(image_data[id], id=id)
+    return children, n_frames_from_image
+
+
+@callback(
+    Output("normalize-psf", "disabled", allow_duplicate=True),
+    Input(dict(type="upload", index="psf"), "last_modified"),
+    prevent_initial_call=True,
+)
+def enable_normalization_button(_):
+    """Re-enable normalization button upon new upload."""
+    return False
+
+
+@callback(
+    Output(dict(type="image-graph", index=MATCH), "figure"),
+    Output(dict(type="image-slider", index=MATCH), "tooltip"),
+    Input(dict(type="image-slider", index=MATCH), "value"),
+    State(dict(type="image-slider", index=MATCH), "tooltip"),
+    State(dict(type="n-frames", index="psf"), "data"),  # NOTE: This intentionally uses index="psf".
+    State("depth-step-size", "value"),
+    State(dict(type="color-scale", index=MATCH), "value"),
+    State(dict(type="image-slider", index=MATCH), "id"),
+    prevent_initial_call=True,
+)
+def update_image_from_slider(value, tooltip, n_frames, step_size, color_scale, id):
```

Comment:
Can you add some type hints here if the types are consistent and reasonable to reference?